src/CachePlugin.php

```diff
@@ -0,0 +1,167 @@
+<?php
+
+namespace Http\Client\Plugin;
+
+use Psr\Cache\CacheItemPoolInterface;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+
+/**
+ * Allow for caching a response.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class CachePlugin implements Plugin
+{
+    /**
+     * @var CacheItemPoolInterface
+     */
+    private $pool;
+
+    /**
+     * Default time to store object in cache. This value is used if CachePlugin::respectCacheHeaders is false or
+     * if cache headers are missing.
+     *
+     * @var int
+     */
+    private $defaultTtl;
+
+    /**
+     * Look at the cache headers to know how long this response is going to be cached.
+     *
+     * @var bool
+     */
+    private $respectCacheHeaders;
+
+    /**
+     * @param CacheItemPoolInterface $pool
+     * @param array                  $options
+     */
+    public function __construct(CacheItemPoolInterface $pool, array $options = array())
+    {
+        $this->pool = $pool;
+        $this->defaultTtl = isset($options['default_ttl']) ? $options['default_ttl'] : null;
+        $this->respectCacheHeaders = isset($options['respect_cache_headers']) ? $options['respect_cache_headers'] : true;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handleRequest(RequestInterface $request, callable $next, callable $first)
+    {
+        $method = strtoupper($request->getMethod());
+
+        // if the request not is cachable, move to $next
+        if ($method !== 'GET' && $method !== 'HEAD') {
+            return $next($request);
+        }
+
+        // If we can cache the request
+        $key = $this->createCacheKey($request);
+        $cacheItem = $this->pool->getItem($key);
+
+        if ($cacheItem->isHit()) {
+            // return cached response
+            return $cacheItem->get();
```

Comment:
Hum, we save a response and so we return a reponse (when hitting the cache), but this method must return a promise, don't think it will work.

Should it not be a `new FulfilledPromise($cacheItem->get())` ?